File: Algorithm/DFS_algorithm.py
#  1st approach
def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)
    print(start)
    for next in graph[start] - visited:
        dfs(graph, next, visited)
    return visited


graph = {'0': set(['1', '2']),
         '1': set(['0', '3', '4']),
         '2': set(['0']),
         '3': set(['1']),
         '4': set(['2', '3'])}
print('1st approach')
print(graph)
dfs(graph,'0')
print()

# 2nd approach
g = {
    'A' : ['B','C'],
    'B' : ['D','E'],
    'C' : ['F'],
    'D' : [],
    'E' : ['F'],
    'F' : []
}

#      A
#     / \
#    B  C
#   / \  \
#  D  E - F

def dfs1(graph, node, visited=None):
    if visited is None:
        visited = set()
    if node not in visited:
        print (node)
        visited.add(node)
        for neighbour in graph[node]:
            dfs1(graph,neighbour,visited)

print('2nd approach')
dfs1(g, 'A')

# 3rd approach

File: Algorithm/DFS_algorithm.java
import java.util.*;

public class DFS_algorithm {

    private static LinkedList<Integer> adjLists[];
    private static boolean visited[];

    public DFS_algorithm(int vertices) {
        adjLists = new LinkedList[vertices];
        visited = new boolean[vertices];

        for(int i=0; i<vertices; i++) {
            adjLists[i] = new LinkedList<Integer>();
        }
    }

    public void addEdge(int src, int dest) {
        adjLists[src].add(dest);
    }

    public void DFS(int vertex) {
        visited[vertex] = true;
        System.out.println(vertex + " ");

        Iterator<Integer> iter = adjLists[vertex].listIterator();
        while(iter.hasNext()) {
            int adj = iter.next();
            if(!visited[adj])
                DFS(adj);
        }
    }
    public static void main(String args[]) {
        DFS_algorithm g = new DFS_algorithm(4);
        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 2);
        g.addEdge(2, 3);

        System.out.println("Following is Depth First Traversal");
        g.DFS(2);
    }
}
